File: src/sdp/brick/Movement.java
/*
 * Copyright (C) 2011 SDP 2011 Group 11
 * This file is part of SDP 2011 Group 11's SDP solution.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with This program.  If not, see <http://www.gnu.org/licenses/>.
 */

import lejos.nxt.Battery;
import lejos.nxt.SensorPort;
import lejos.nxt.TouchSensor;

abstract class OrderNode {
	private final static SimpleMotor left = SimpleMotor.B;
	private final static SimpleMotor right = SimpleMotor.C;
	private boolean sharpTurn = false;
	
	public abstract boolean execute(Location state) throws InterruptedException;
	
	protected void brake() throws InterruptedException {
		left.stop();
		right.stop();
	}
	
	protected void idle() {
		left.idle();
		right.idle();
	}
	
	protected void changePower(int leftPower, int rightPower, boolean goBackwards) {
		if(goBackwards) {
			left.run(-rightPower);
			right.run(-leftPower);
		} else {
			left.run(leftPower);
			right.run(rightPower);
		}
		
		if(Math.abs(leftPower - rightPower) <= 5)
			sharpTurn = false;
		else if((leftPower < 0) != (rightPower < 0))
			sharpTurn = true;
		else {
			int diff = Math.abs(leftPower - rightPower);
			int ma = Math.max(Math.abs(leftPower), Math.abs(rightPower));
			sharpTurn = diff / (float) ma > 0.15f;
		}
	}
	
	protected static int scalePower(int power) {
		return Math.max(-100, Math.min(100, (int) Math.round(power * Movement.powerMultiplier * 1.3)));
	}
	
	public boolean turning() {
		return sharpTurn;
	}
}

class MovementNode extends OrderNode {
	private static final float[] pastAngle = new float[50];
	private static int anglePos = 0;
	private static int lastPower = 0;
	private static long lastPowerChange = 0;
	
	static {
		for(int i = 0; i < pastAngle.length; ++i)
			pastAngle[i] = 5;
	}
	
	private static boolean stable() {
		for(int i = 0; i < pastAngle.length; ++i)
			if(!(pastAngle[i] < 0.2 || pastAngle[i] > 6.1))
				return false;
		return true;
	}
	
	private float x, y;
	private boolean stop, allowReversing, turnHard = false;
	
	MovementNode(float x, float y, boolean stop, boolean allowReversing) {
		this.x = x;
		this.y = y;
		this.stop = stop;
		this.allowReversing = allowReversing;
		
		long now = System.currentTimeMillis();
		if(lastPower < scalePower(Movement.initStraightPower) || now - lastPowerChange > 1000) {
			lastPower = scalePower(Movement.initStraightPower);
			lastPowerChange = System.currentTimeMillis();
		}
	}
	
	public boolean execute(Location state) {
		float dist = (float) Math.sqrt(Math.pow((state.x - this.x), 2.0) + Math.pow(state.y - this.y, 2.0));
		if(stop) {
			if(dist < 2)
				return true;
		} else {
			if(dist < 3)
				return true;
		}
		
		final long now = System.currentTimeMillis();
		
		float targetRotation = Utility.normalizeRotation((float) Math.atan2(this.x - state.x, this.y - state.y));
		float angle = Utility.normalizeRotation(targetRotation - state.rotation);
		float difference = (float) Math.min(angle, 2 * Math.PI - angle);
		boolean backwardsMovement = false;
		
		if(((state.x - this.x > 3) || allowReversing) && difference > (3 * Math.PI) / 4) {
			angle = Utility.normalizeRotation((float) (angle + Math.PI));
			difference = (float) Math.min(angle, 2 * Math.PI - angle);
			backwardsMovement = true;
		}
		
		pastAngle[anglePos] = angle;
		++anglePos;
		if(anglePos == pastAngle.length)
			anglePos = 0;
		
		if(turnHard || difference > Movement.maxSoftTurn) {
			if(difference < 0.05) {
				turnHard = false;
				return execute(state);
			}
			final int power = scalePower(Movement.initMidStraightTurnPower);
			turnHard = true;
			lastPowerChange = now;
			lastPower = power;
			if(angle < Math.PI)
				changePower(power, -power, backwardsMovement);
			else
				changePower(-power, power, backwardsMovement);
		} else {
			lastPower = Math.max(lastPower, scalePower(Movement.initStraightPower));
			int maxPower = (int) Math.round(scalePower(Movement.initStraightPower) + Math.max(0, dist - Movement.powerLimitZone)
					* Movement.maxPowerLimiterConstant);
			maxPower = Math.max(scalePower(Movement.initStraightPower), Math.min(100, maxPower));
			
			if(stable()) {
				if(now - lastPowerChange >= Movement.accelerationInterval && lastPower + 1 <= maxPower) {
					lastPowerChange = now;
					++lastPower;
				} else if(lastPower > maxPower) {
					lastPowerChange = now;
					lastPower = maxPower;
				}
			} else {
				lastPowerChange = now;
				lastPower = scalePower(Movement.initStraightPower);
			}
			
			// slow down one side only
			final int diff = (int) Math.floor(lastPower * Movement.maxHeadingStabilizer * difference / Movement.maxSoftTurn);
			
			if(angle < Math.PI)
				changePower(lastPower, lastPower - diff, backwardsMovement);
			else
				changePower(lastPower - diff, lastPower, backwardsMovement);
		}
		
		state.sharpTurn = turning();
		return false;
	}
}

class TurningNode extends OrderNode {
	private final float rotation;
	
	TurningNode(float rotation) {
		this.rotation = rotation;
	}
	
	public boolean execute(Location state) {
		float angle = Utility.normalizeRotation(this.rotation - state.rotation);
		float difference = Math.min(angle, (float) (2 * Math.PI - angle));
		if(difference < 0.05)
			return true;
		
		int power = scalePower(Movement.initTurnPower);
		if(angle < Math.PI)
			changePower(power, -power, false);
		else
			changePower(-power, power, false);
		
		state.sharpTurn = turning();
		return false;
	}
}

class StopNode extends OrderNode {
	protected boolean done = false;
	
	public boolean execute(Location state) throws InterruptedException {
		state.sharpTurn = false;
		if(done)
			return true;
		brake();
		done = true;
		return false;
	}
}

public class Movement extends Thread {
	static float powerMultiplier;
	static float maxSoftTurn;
	static float maxHeadingStabilizer;
	static float maxPowerLimiterConstant;
	static int powerLimitZone;
	static int accelerationInterval;
	static int initStraightPower;
	static int initMidStraightTurnPower;
	static int initTurnPower;
	
	private static void updateConstants() {
		powerMultiplier = DataContainer.powerMultiplier;
		maxSoftTurn = DataContainer.maxSoftTurn;
		maxHeadingStabilizer = DataContainer.maxHeadingStabilizer;
		maxPowerLimiterConstant = DataContainer.maxPowerLimiterConstant;
		powerLimitZone = DataContainer.powerLimitZone;
		accelerationInterval = DataContainer.accelerationInterval;
		initStraightPower = DataContainer.initStraightPower;
		initMidStraightTurnPower = DataContainer.initMidStraightTurnPower;
		initTurnPower = DataContainer.initTurnPower;
	}
	
	private static TouchSensor touchFrontLeft = new TouchSensor(SensorPort.S1);
	private static TouchSensor touchFrontRight = new TouchSensor(SensorPort.S2);
	private static TouchSensor touchRearLeft = new TouchSensor(SensorPort.S3);
	private static TouchSensor touchRearRight = new TouchSensor(SensorPort.S4);
	private static SimpleMotor left = SimpleMotor.B;
	private static SimpleMotor right = SimpleMotor.C;
	
	private final Location state;
	private int orderNumber = 0;
	private int touchSpeed = 65;
	private MoveCommand cmd = null;
	private OrderNode cur = null;
	private int movePos = 0;
	
	public Movement(Location state) {
		super();
		this.state = state;
	}
	
	private boolean handleTouch() throws InterruptedException {
		if(!touchFrontLeft.isPressed() && !touchFrontRight.isPressed() && !touchRearLeft.isPressed() && !touchRearRight.isPressed())
			return false;
		Thread.sleep(50);
		
		if((touchFrontLeft.isPressed() || touchFrontRight.isPressed()) && (touchRearLeft.isPressed() || touchRearRight.isPressed())) {
			left.stop();
			right.stop();
		} else if(touchFrontLeft.isPressed() && touchFrontRight.isPressed()) {
			left.run(-touchSpeed);
			right.run(-touchSpeed);
		} else if(touchFrontLeft.isPressed()) {
			left.run(touchSpeed);
			right.run(-touchSpeed);
		} else if(touchFrontRight.isPressed()) {
			left.run(-touchSpeed);
			right.run(touchSpeed);
		} else if(touchRearLeft.isPressed() && touchRearRight.isPressed()) {
			left.run(touchSpeed);
			right.run(touchSpeed);
		} else if(touchRearLeft.isPressed()) {
			left.run(touchSpeed);
			right.run(-touchSpeed);
		} else if(touchRearRight.isPressed()) {
			left.run(-touchSpeed);
			right.run(touchSpeed);
		}
		Thread.sleep(50);
		return true;
	}
	
	private void loadNextNode() {
		if(movePos < cmd.getRoute().size()) {
			cur = new MovementNode(cmd.getRoute().get(movePos).x, cmd.getRoute().get(movePos).y, movePos + 1 == cmd.getRoute().size(), cmd.reversingAllowed());
			++movePos;
		} else if(cmd.getRoute().size() == movePos) {
			if(cmd.getFinalRotation() != null)
				cur = new TurningNode(cmd.getFinalRotation());
			else
				cur = new StopNode();
			++movePos;
		} else {
			if(cur == null || !(cur instanceof StopNode))
				cur = new StopNode();
		}
	}
	
	public void run() {
		final VisionLocationTracker vision = new VisionLocationTracker(state);
		final LocationTracker tracker = new LocationTracker(state);
		
		int lastFrequencySent = (int) System.currentTimeMillis(), lastVoltageSent = 0, lastTouch = 0;
		int iterationCounter = 0, now;
		boolean inTouchState = false;
		
		try {
			while(true) {
				now = (int) System.currentTimeMillis();
				++iterationCounter;
				
				if(now - lastVoltageSent >= 30000) {
					if(BrickSender.haveConnection())
						BrickSender.add("/brick/info/battery " + Battery.getVoltage());
					lastVoltageSent = now;
				}
				
				if(now - lastFrequencySent >= 500) {
					if(BrickSender.haveConnection())
						BrickSender.add("/brick/info/frequency " + (iterationCounter * 1000 / (now - lastFrequencySent)));
					lastFrequencySent = now;
					iterationCounter = 0;
				}
				
				if(DataContainer.move != null) {
					synchronized(DataContainer.lock) {
						cmd = DataContainer.move;
						DataContainer.move = null;
						updateConstants();
					}
					movePos = 0;
					loadNextNode();
					++orderNumber;
				}
				
				tracker.updateLocation();
				vision.updateLocation(state);
				
				if(handleTouch()) {
					inTouchState = true;
					lastTouch = now;
					continue;
				} else if(inTouchState) {
					if(now - lastTouch <= 400) {
						Thread.sleep(1);
						continue;
					}
					left.stop();
					right.stop();
					if(now - lastTouch <= 500) {
						Thread.sleep(1);
						continue;
					}
					inTouchState = false;
					continue;
				}
				
				if(cur.execute(state))
					loadNextNode();
				
				Thread.sleep(1);
			}
		} catch(InterruptedException e) {
			return;
		}
	}
}

File: src/sdp/movement.py
# ###################################################
# Copyright (C) 2011 SDP 2011 Group 11
# This file is part of SDP 2011 Group 11's SDP solution.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with This program.  If not, see <http://www.gnu.org/licenses/>.
# ###################################################

from __future__ import with_statement
import __init__ # uses psyco if available @UnusedImport
from datetime import datetime
import threading
import copy
import math
import time
import sys
import re

START_TIME = datetime.now().strftime('%Y-%m-%dT%H:%M:%S.%f')
print '/general/identify movement primary ' + START_TIME + ' simple movement module'
sys.stdout.flush()

def normalize_rotation(rotation):
    full_circle = 2 * math.pi
    while rotation < 0:
        rotation += full_circle
    while rotation > full_circle:
        rotation -= full_circle
    return rotation

class WorldSituation(object):
    def __init__(self, x, y, rotation):
        self.x = x
        self.y = y
        self.rotation = rotation
    
    def known(self):
        return self.x != None and self.y != None and self.rotation != None

    def __str__(self):
        return 'WorldSituation: ' + str(self.x) + ', ' + str(self.y) + ' (' + str(self.rotation) + ')'

class OrderNode(object):
    def __init__(self):
        self._last_left = ''
        self._last_right = ''
    
    def _send_msg(self, do_left, do_right):
        any = False
        if self._last_left != do_left:
            self._last_left = do_left
            print do_left
            any = True
        if self._last_right != do_right:
            self._last_right = do_right
            print do_right
            any = True
        if any:
            sys.stdout.flush()
    
    def _change_speed(self, left, right):
        if left == 0:
            do_left = '/movement/order/idle left'
        else:
            do_left = '/movement/order/run_inf left ' + str(left)
        if right == 0:
            do_right = '/movement/order/idle right'
        else:
            do_right = '/movement/order/run_inf right ' + str(right)
        self._send_msg(do_left, do_right)
    
    def _brake(self):
        self._send_msg('/movement/order/brake left', '/movement/order/brake right')
    
    def _idle(self):
        self._send_msg('/movement/order/idle left', '/movement/order/idle right')

def scale_speed(speed):
    speed = int(round(speed * 10))
    if speed < -128:
        speed = -128
    if speed > 127:
        speed = 127
    if speed < 0:
        if speed > -30:
            speed = -30
    if speed > 0:
        if speed < 30:
            speed = 30
    return speed

class MovementNode(OrderNode):
    def __init__(self, x, y, stop):
        OrderNode.__init__(self)
        self.x = x
        self.y = y
        self.stop = stop
    
    # returns True if the work has been done
    def execute(self, state):
        if not state.known():
            return False
        
        dist = math.sqrt((state.x - self.x) ** 2 + (state.y - self.y) ** 2)
        if dist < 3:
            return True
        speed = scale_speed(25)
        if dist > 70:
            speed = scale_speed(40)
        elif dist > 40:
            speed = scale_speed(30)
        
        target_rotation = normalize_rotation(math.atan2(self.x - state.x, self.y - state.y))
        angle = normalize_rotation(target_rotation - state.rotation)
        difference = min(angle, 2 * math.pi - angle)
        max_diff = 0.3
        
        if difference > max_diff:
            #speed = min(speed, scale_speed(25))
            if angle < math.pi:
                self._change_speed(speed, -speed)
            else:
                self._change_speed(-speed, speed)
            if difference < 0.1:
                time.sleep(0.04)
                self._idle()
        else:
            max_diff_percent = 0.2
            diff = int(math.floor(speed * max_diff_percent * difference / max_diff))
            
            if angle < math.pi:
                self._change_speed(speed, speed - diff)
            else:
                self._change_speed(speed - diff, speed)
            
            if dist < 8 and self.stop:
                time.sleep(0.04)
                self._idle()
        
        return False
    
    def __str__(self):
        return 'MovementNode: ' + str(self.x) + ', ' + str(self.y) + (' (final move)' if self.stop else ' (not final move)')

class TurningNode(OrderNode):
    def __init__(self, rotation):
        OrderNode.__init__(self)
        self.rotation = rotation
    
    # returns True if the work has been done
    def execute(self, state):
        if not state.known():
            return False
        
        angle = normalize_rotation(self.rotation - state.rotation)
        difference = min(angle, 2 * math.pi - angle)
        if difference < 0.05:
            return True
        
        speed = scale_speed(25)
        if angle < math.pi:
            self._change_speed(speed, -speed)
        else:
            self._change_speed(-speed, speed)
        
        if difference < 0.5:
            time.sleep(0.04)
            self._idle()
        return False
    
    def __str__(self):
        return 'TurningNode: ' + str(self.rotation)

class StopNode(OrderNode):
    def __init__(self):
        OrderNode.__init__(self)
        self._done = False
    
    # returns True if the work has been done
    def execute(self, state):
        if self._done:
            return True
        self._brake()
        self._done = True
        return False
    
    def __str__(self):
        if self._done:
            return 'StopNode: done'
        return 'StopNode: not done'

class MovementOrder(object):
    def __init__(self, cmd):
        # /planner/move n coordinates (final_rotation|not_important)
        pieces = cmd.split(' ')
        n = int(pieces[0])
        self._parts = []
        for i in range(n):
            self._parts.append(MovementNode(float(pieces[2 * i + 1]), float(pieces[2 * i + 2]), i + 1 == n))
        rotation = pieces[len(pieces) - 1]
        if rotation != 'not_important':
            self._parts.append(TurningNode(float(rotation)))
        self._parts.append(StopNode())
        self._pos = 0
    
    def move(self, state):
        # passes control to the next order until it runs out or finds one that needs to do work
        while self._pos < len(self._parts) and self._parts[self._pos].execute(state):
            self._pos += 1
    
    def __str__(self):
        res = 'MovementOrder at ' + str(self._pos) + '/' + str(len(self._parts))
        for node in self._parts:
            res += '\n' + str(node)
        return res

class StreamListener(threading.Thread):
    def __init__(self):
        threading.Thread.__init__(self, name = 'stream listener')
        self.daemon = True
        self._lock = threading.Lock()
        self._state = WorldSituation(None, None, None)
        self._cmd = None
    
    def run(self):
        time_format = '([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{2}):([0-9]{2}):([0-9]{2})\.([0-9]+)' # 7 groups
        real_format = '-?[0-9]+(\.[0-9]+)?([eE]-?[0-9]+)?' # 2 groups
        regex_vision_xy = re.compile('/vision/xy\s+us\s+(' + real_format + ')\s+(' + real_format + ')\s+' + time_format + '$')
        regex_vision_rotation = re.compile('/vision/rotation\s+us\s+(' + real_format + ')\s+' + time_format + '$')
        planner_move_prefix = '/planner/move '
        planner_move_prefix2 = '/planner/move_rev '
        
        while True:
            msg = sys.stdin.readline().rstrip()
            if msg == '/central/exit':
                print '/general/exiting obeying exit message'
                sys.stdout.flush()
                break
            elif msg == '':
                print '/general/exiting received empty message'
                sys.stdout.flush()
                break
            else:
                with self._lock:
                    match = regex_vision_xy.match(msg)
                    if(match != None):
                        self._state.x = float(match.group(1))
                        self._state.y = float(match.group(4))
                        continue
                    
                    match = regex_vision_rotation.match(msg)
                    if(match != None):
                        self._state.rotation = float(match.group(1))
                        continue
                    
                    if msg[0: len(planner_move_prefix)] == planner_move_prefix:
                        self._cmd = msg[len(planner_move_prefix):].strip()
                        continue
                    elif msg[0: len(planner_move_prefix2)] == planner_move_prefix2:
                        self._cmd = msg[len(planner_move_prefix2):].strip()
                        continue
    
    def get_state(self):
        with self._lock:
            return copy.deepcopy(self._state)
    
    def take_cmd(self):
        with self._lock:
            t = copy.deepcopy(self._cmd)
            self._cmd = None
            return t

if __name__ == "__main__":
    stream_listener = StreamListener()
    stream_listener.start()
    order = MovementOrder('0 not_important')
    old_state = ''
    
    while True:
        state = stream_listener.get_state()
        cmd = stream_listener.take_cmd()
        if cmd != None:
            order = MovementOrder(cmd)
        elif str(state) == old_state:
            time.sleep(0.001)
            continue
        
        old_state = str(state)
        order.move(state)
        
        time.sleep(0.001)